### ISSUE TITLE
Implementation feedback

### ISSUE THREAD

**Issue by samuel-lucas6** [jedisct1/draft-aegis-aead#61]
This is a summary of feedback from @fscoto implementing AEGIS-256. Hopefully you don't mind me putting this here for tracking purposes. Well done by the way :)

1. ~Duplicate `C0` and `C1` in the `Init()` sections?~
2. ~Rename `Pad()` to `ZeroPad()`?~
3. Full annotated execution of `Encrypt()` for an ad/message between 17 and 31 bytes in length?
4. ~Include zeroing `msg` check in a test vector?~
5. ~Consider having a `key` that isn't all zeroes in the first test vector?~
6. ~`C_MAX` makes little sense given the tag is a separate input.~
7. ~`DecPartial()` but no `EncPartial()` feels weird.~
8. ~Maybe leave `msg` out of test vectors 6-9 to make it clear that only `Decrypt()` is being tested?~

**Comment by samuel-lucas6** [comment]
1. I would find this annoying as well, but you can make the same argument for most of the conventions and definitions, so perhaps not.
2. This seems reasonable for slightly more clarity.
3. Please explain what you mean.
4. I agree given the Security Considerations section.
5. I'll leave that to you two.
6. That is odd and should be changed.
7. Yes, but you're right that `EncPartial()` would be small. If you think it would help, add it.
8. That makes sense.

**Comment by jedisct1** [comment]
1. That would *decrease* readability IMHO.
2. Good idea.
3. That would be having a step-by-step example of encrypting a full block followed by a partial block. Why not. But given the state size, that would add a significant amount of content to the document, though.
4. I'm not totally convinced, as the recommendation only makes sense for C implementations. Virtually all modern languages would return an exception or an error, and callers can't even access the decrypted message except when doing in-place decryption.
5. The test vectors with an all-zero key comes from the original paper. But having a key that places different content in every register can indeed by very useful to quickly spot implementation errors. So, yes, big +1 here even if it means changing the original test vectors.
6. That was assuming that the tag was attached to the ciphertext, but yeah, it's a bit weird, a we should just define is as `P_MAX`.
7. Partial block decryption is tricky, while partial block encryption just adds `ZeroPad()` call to full block encryption. Defining and calling a function would be more verbose and I'm not sure that any real-world implementation would actually use define a dedicated function for this.

**Comment by fscoto** [comment]
> Full annotated execution of `Encrypt()` for an ad/message between 17 and 31 bytes in length? (Not sure what this means)
>
> Please explain what you mean.

I'm suggesting Dump something like this into an appendix (similarly for -128L):

```
key  : 000102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f
nonce: 202122232425262728292a2b2c2d2e2f303132333435363738393a3b3c3d3e3f
ad   : 404142434445464748494a4b4c4d4e4f50
msg  : 606162636465666768696a6b6c6d6e6f7071
Initial state
S0: 20202020202020202020202020202020
S1: 20202020202020202020202020202020
S2: db3d18556dc22ff12011314273b528dd
S3: 000101020305080d1522375990e97962
S4: 0000030107000e0a1d2b3d529ce4776d
S5: cb2c0a4679d739e638082b596fa836c2
After Init() update 1
S0: 86b023c7bdded5ba32b903df5de70a4b
S1: 31a95a24c20ece6c4be7f3cee4aacd5e
S2: 7f87b99d0b3c89d3aba838b8873404b0
S3: 62241d55e56578edc2ba514463d09a03
S4: d9f885d1270c3f9f641a6d1a31ce5b5d
S5: d988593411aec838e94de574d8b2e045
After Init() update 2
S0: 3bb10235f5c5b923ff881cdeac57c12a
S1: c76bcfb6ff434eb3fc18bc9e2f31638e
S2: f9cb4b6e61b95b5beb8339c175a86c41
S3: 26905d50a0a999bd1006884e93cc326e
S4: 77fd205fabe2006131e1438751fd0e3a
S5: c307051e6bcd8e7105cde004eed770b8
After Init() update 3
S0: 87408e38f5b273a7828c823d87a49a4e
S1: ab7ec22fc6f780cb72ead0d292bfd6c9
S2: 6fc27ea6eb9cf59a3ce58f38dafcd74c
S3: e815f7b3d7da902766d8a8162c0ddaef
S4: 7264b47e236acdd144cca3088291a02f
S5: 9b3df9838976e41bdb66af9dc78b13a8
After Init() update 4
S0: a3f0eb54deec2c7397ba891e07dba256
S1: 02c9e554a573c761d305510b068a513b
S2: 8284d4ac9c73cba814b270e5b20f9803
S3: d592a9ff88e922497fcfd19aa06f400c
S4: 90dbb3444d3c7949f1b6d262a22b8984
S5: 1c3ad63929b62bc31915ef0b16dfd569
Processing ad
After Update in Enc(404142434445464748494a4b4c4d4e4f)
S0: 3432ced031a9a6be0d2f0c0684e9aff7
S1: 4907bc5a82f33ac86e4b23dd1bad6c76
S2: d57ca8b4dceb7f586cb7eb2dfdbbc0ff
S3: 530c1ae2a102ef7880c3902ee91e66bc
S4: 7458cb8644fa1fb1457a58028343c808
S5: 100ccd799a20fe221ec12819aee7c467
After Update in Enc(50000000000000000000000000000000)
S0: 98544f0778739f7e6021258421e8bf36
S1: 81d3288f21e58a16dc880e6d8bb7e94e
S2: aa0f9aa2dacba6e9605e08acaddfe9c9
S3: 8af003cb619523fbad6b0b227bdfee29
S4: 299e3e77f959d6d14c1267fd48f1de69
S5: 024af6f375f338b53b72212564f3400c
Processing msg
After Update in Enc(606162636465666768696a6b6c6d6e6f)
S0: 02baafeaf91d21f657a3402c92470b33
S1: bdd490ff93658fada61158e48e237817
S2: 4663739a01f436c9e48a862e3f900aee
S3: 05efacaa036a5e3009510c53655bede6
S4: 256cc8a58a2f3792b0228ba2302a919d
S5: 690f99a0f43156201df194317fe212af
ci: 606162636465666768696a6b6c6d6e6f
After Update in Enc(70710000000000000000000000000000)
S0: 92b85eed4d5b8d9feeb279d8c176840c
S1: 6e280f560812d829d0d897549bf2b12d
S2: ffcd9bc895fe31a433d7e035cfe5062c
S3: 0734b9efa6fb41c629af7ca6adfd3ae5
S4: 8594cfe3edbed8d86cfd860165c79f0c
S5: 4b032b7521eca8bc5bab8c9027c9271d
ci: 85a5e170ec1bf81f0bc24375e4fbf3c3
Finalize(0x88, 0x90)
t:  8f34b9efa6fb41c6b9af7ca6adfd3ae5
After Update in Finalize()
S0: 8c4b39274d38b59ed185f266782faae3
S1: a825f73b15796dd96b1e84553ff336a4
S2: 37be585a416826e0dea9bc91a3300523
S3: dfcf62919af6eafe98b380c910eba9b1
S4: b7c9c65e0c18d940e427545a5ccf29b4
S5: 1af675a2762891e665acb2538baa544f
tag:6120472bf9ef5cff7d0aac6223624d2e
```

Points 2 and 8 already done thanks to @samuel-lucas6.

Points 1, 3, 4, 7 dismissed or no clear consensus. Nothing to do.

5 still to do.

6 still to do, might actually be correct as-is, cf. [RFC 5116, § 5.1](https://www.rfc-editor.org/rfc/rfc5116#section-5.1), where tag size is incorporated, but also mandates ct || tag, but § 4 should leave us with enough leeway.

**Comment by samuel-lucas6** [comment]
> Upon checking the draft again, the C_MAX definitions seem correct. In an AEAD, the tag is part of what is transmitted over the network. We correctly note that C_MAX is just P_MAX plus tag length. Nothing to do for 6.

I agree that makes sense, but it seems wrong to have `tag` listed as a separate input if it's actually part of the ciphertext. Two options I see are:
1. You only have a `ct` input, like implementations presumably would, but then you'd have to specify where `tag` is coming from somehow.
2. You say `ct` length **MUST** be less than `P_MAX` and treat it as independent from `tag`.

> Unless someone has strong feelings about doing 3, we've addressed all we can agree to address and the issue may be closed.

Given the length, perhaps we should leave it out.

**Comment by jedisct1** [comment]
The confusion comes from the fact that there's no dedicated parameter for the tag size.

Maybe what we can do is keep the current definition for `C_MAX` and explicitly mention "(incorporating the authentication tag)"?

For 3), maybe what we can do is provide a single test vector (for each AEGIS variant) for the `Update()` function?

**Comment by fscoto** [comment]
> Maybe what we can do is keep the current definition for C_MAX and explicitly mention "(incorporating the authentication tag)"?

Works for me in principle, though I'm having issues finding a good wording for it.

> For 3), maybe what we can do is provide a single test vector (for each AEGIS variant) for the Update() function?

We already have that. While that was indeed helpful, the part I wasn't sure is if my state was okay after Init(), after processing ad and after processing msg. I had a bug in init() and multi-block ad processing, so ct and msg always came out wrong. Without being sure that ad processing works, it was hard to tell if the bug was in ad or msg handling. The Init() bug is caught with the new test vectors.

It's probably not a big deal since if test vector 1 passes but 3 doesn't, completely hosed ct means bug in ad handling and partially hosed ct means bug in msg handling.

**Comment by samuel-lucas6** [comment]
I've now implemented AEGIS-128L and AEGIS-256 in C#. I used a mixture of the draft and Zig reference implementation for AEGIS-256 and only the draft for AEGIS-128L. Neither worked first try due to a draft misread with the former and a C# API issue with the latter. Here's my feedback:

1. I think the description for `LE64(x)` should say `unsigned integer x` for clarity.
2. The constants (`C0` and `C1`) were annoying to convert to bytes. I copied them from the reference implementation in the end, but `c0` appears to be missing a digit for the first two (`0x0, 0x1`).
3. `nonce` was described as `the public nonce` sometimes and `the nonce` other times. It should be consistent.
4. I didn't realise `ad_len` and `msg_len` were in bits, which led to the tests failing. It would be clearer having them as bytes and including `* 8` in the pseudocode. `tag_length == 16` is in bytes.
5. The comma can be removed from `Either the decrypted message msg, or an error indicating that the authentication tag is invalid for the given inputs` in 3.2 and 4.2.
6. `It returns the 128-bit block out.` can be removed from 4.6. Think this was left over after a previous edit.
7. Something might want to be added to the Introduction about where 256-bit tag support came from.

**Comment by jedisct1** [comment]
This is very useful feedback, thank you Samuel1

The constants in the reference implementations have been changed to add leading zeros. Obviously, that doesn't change their value, but it indeed makes them consistent with how other values are encoded.

We can provide these constants as an array in the specification if it makes implementations easier. This is a good idea.

`nonce` is described as `public nonce` to emphasize the fact that it's public. There are discussions and proposals about secret/encrypted nonces, so this information should be mentioned. But maybe only once.

It's very common for ciphers and hash functions to specify the length in bits. This is very likely due to NIST requirements, but I can still see that being useful, especially with languages supporting arbitrary-long types.

As an alternative to enforcing a multiple of `8`, how about renaming `ad_len` and `msg_len` to `ad_len_bits` and `msg_len_bits`?

**Comment by samuel-lucas6** [comment]
> We can provide these constants as an array in the specification if it makes implementations easier.

I definitely think that's worth doing.

> nonce is described as public nonce to emphasize the fact that it's public. There are discussions and proposals about secret/encrypted nonces, so this information should be mentioned. But maybe only once.

Yes, that makes sense. I only noticed it because `public` is missing from the Init function sections (3.3 and 4.3), whereas `key: the encryption key` is consistent throughout. I'd say either `public` should be kept in the Conventions section and removed elsewhere, or it should be added to the Init sections.

> As an alternative to enforcing a multiple of 8, how about renaming ad_len and msg_len to ad_len_bits and msg_len_bits?

Sounds good. Looking again, it was already quite clear if reading properly, but this sort of change doesn't hurt.

Perhaps `tag_length == 16` should be in bits too for consistency, although I like the way it's commented currently.

**Comment by samuel-lucas6** [comment]
Closing this after previous work addressed Fabio's feedback and now #78 addressed mine.
